drivers: modem: Adding SMS messaging support

johnlange2 opened the pull request:
Adding function pointers for sending and receiving SMS messages
to the modem context. The receive SMS function may be blocking.

Signed-off-by: John Lange <John.Lange2@T-Mobile.com>

johnlange2 commented:
**Documentation for SMS Messaging Support in Zephyr**

Background: SMS (Short Message Service) is defined in RFC 5724

Below are some notes about the implementation:
-a new file, modem_sms.h, defines SMS sizes, structs, and functions
-two new function members were added to struct modem_context
-SMS support is modem-specific (must be implemented per modem)
-SMS receive is not guaranteed to return the oldest message first
-SMS receive is non-blocking if timeout is K_NO_WAIT
-SMS send/receive can be invoked using a modem context or a socket

Below are some examples:

Example 1: Send SMS message using modem context

```
#include <stddef.h>
#include <errno.h>
#include "modem_context.h"

int modem_send_sms(const int modem_index, const struct sms_out *sms)
{
	struct modem_context *ctx = modem_context_from_id(modem_index);

	if (ctx->send_sms == NULL) {
		return -ENOSYS;
	}

	return ctx->send_sms(ctx, sms);
}
```

Example 2: Receive SMS message using modem context

```
#include <stddef.h>
#include <errno.h>
#include "modem_context.h"

int modem_recv_sms(const int modem_index, struct sms_in *sms)
{
	struct modem_context *ctx = modem_context_from_id(modem_index);

	if (ctx->recv_sms == NULL) {
		return -ENOSYS;
	}

	return ctx->recv_sms(ctx, sms, K_NO_WAIT);
}
```

Example 3: Send SMS message using a socket

```
#include <fcntl.h>
#include "modem_sms.h"

int sms_send_on_socket(const int sock_fd, const struct sms_out *sms)
{
	// Call fcntl to send the SMS message
	return fcntl(sock_fd, SMS_SEND, &sms);
}

```
Example 4: Receive SMS message using a socket

```
#include <fcntl.h>
#include "modem_sms.h"

int sms_receive_on_socket(const int sock_fd, struct sms_in *sms)
{
	// Call fcntl to recv an SMS message
	return fcntl(sock_fd, SMS_RECV, sms, K_NO_WAIT);
}
```

rerickson1 reviewed (dismissed):
Looks good!

jukkar commented on drivers/modem/modem_context.h:
```diff
@@ -19,15 +20,16 @@
 
 #define MODEM_PIN(name_, pin_, flags_) { \
-	.dev_name = name_, \
-	.pin = pin_, \
-	.init_flags = flags_ \
+		.dev_name = name_,	 \
+		.pin = pin_,		 \
+		.init_flags = flags_	 \
```
The whitespace changes here and below are fine but should be placed in separate commit as they are not really related to SMS changes.

jukkar commented on drivers/modem/modem_context.h:
```diff
@@ -67,12 +69,12 @@ struct modem_context {
 #if defined(CONFIG_MODEM_CELL_INFO)
-	int   data_operator;
-	int   data_lac;
-	int   data_cellid;
+	int data_operator;
+	int data_lac;
+	int data_cellid;
 #endif
 	int   *data_rssi;
-	bool  is_automatic_oper;
+	bool is_automatic_oper;
```
ditto

johnlange2 replied:
Done

jukkar commented on drivers/modem/modem_context.h:
```diff
@@ -83,6 +85,10 @@ struct modem_context {
 	struct modem_cmd_handler cmd_handler;
 
+	/* SMS functions */
+	send_sms_func send_sms;
+	recv_sms_func recv_sms;
```
The naming of these function pointers in other part of the system seems to be `xxxx_func_t` so I propose we follow the same naming convention.

jukkar commented on drivers/modem/modem_shell.c:
```diff
@@ -242,11 +245,114 @@ static int cmd_modem_info(const struct shell *shell, size_t argc, char *argv[])
+	}
+	else {
+		int modem_index = atoi(argv[1]);
+		ms_ctx = ms_context_from_id(modem_index);
```
Please put one empty line after variable definitions (to be consistent with other part of the system)

johnlange2 replied:
Done

jukkar commented on drivers/modem/modem_shell.c:
```diff
@@ -242,11 +245,114 @@ static int cmd_modem_info(const struct shell *shell, size_t argc, char *argv[])
+		if (ms_ctx == NULL) {
+			shell_fprintf(shell, SHELL_ERROR,
+				      "Couldn't open context for modem index %d\n", modem_index);
+			return -1;
```
At least in the network shell, the `-ENOEXEC` is returned in case of an error. Don't know if it makes much difference here.

jukkar commented on drivers/modem/modem_shell.c:
```diff
@@ -242,11 +245,114 @@ static int cmd_modem_info(const struct shell *shell, size_t argc, char *argv[])
+			return -1;
+		}
+
+		snprintf(sms.phone, sizeof(sms.phone), "%s", argv[2]);
+		snprintf(sms.msg, sizeof(sms.msg), "%s", argv[3]);
```
I propose `snprintk` instead, then we do not need to include stdio.h and I recall that the snprintk will also use less memory

jukkar commented on drivers/modem/modem_shell.c:
```diff
@@ -242,11 +245,114 @@ static int cmd_modem_info(const struct shell *shell, size_t argc, char *argv[])
+		snprintf(sms.msg, sizeof(sms.msg), "%s", argv[3]);
+
+		ret = ms_ctx->send_sms(ms_ctx, &sms);
+
```
Empty line here can be removed, our coding style puts variable setting and the corresponding `if` statement together.

jukkar commented on drivers/modem/modem_shell.c:
```diff
@@ -242,11 +245,114 @@ static int cmd_modem_info(const struct shell *shell, size_t argc, char *argv[])
+			      "usage: modem sms recv <modem index>\n");
+	}
+	else {
+		int modem_index = atoi(argv[1]);
```
empty line after variable declaration

johnlange2 replied:
Done

jukkar commented on drivers/modem/modem_shell.c:
```diff
@@ -242,11 +245,114 @@ static int cmd_modem_info(const struct shell *shell, size_t argc, char *argv[])
+		}
+
+		ret = ms_ctx->recv_sms(ms_ctx, &sms, K_NO_WAIT);
+
```
remove empty line

jukkar commented on drivers/modem/modem_context.h:
```diff
@@ -7,6 +7,7 @@
 
```
The shell changes should be placed in a separate commit in this PR

johnlange2 replied:
Done

jukkar commented on drivers/modem/modem_sms.h:
```diff
@@ -0,0 +1,47 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
```
What is the purpose of this pragma, why do we need it?

johnlange2 replied:
From [https://en.wikipedia.org/wiki/Pragma_once](https://en.wikipedia.org/wiki/Pragma_once), "In the C and C++ programming languages, pragma once is a non-standard but widely supported preprocessor directive designed to cause the current source file to be included only once in a single compilation". Basically, it is used instead of the three lines: #ifndef and #define at the top, and #endif at the bottom.

jukkar replied:
Please, do not do that. Just put ifndef+define+endif like the other include files in the system.

jukkar reviewed (changes requested):
Minor comments added. Could you create a sample app that shows how the API is used? I know there is the shell that shows this too but typically we have a sample app for that purpose as then it gets included in the documentation etc.

johnlange2 committed:
drivers: modem: Adding SMS support to modem shell
Adding menu and support functions for sending and receiving
SMS messages to the modem shell

Signed-off-by: John Lange <john@y2038.com>

johnlange2 committed:
drivers: modem: Adding header file for SMS messaging
Adding header file for SMS messaging support

Signed-off-by: John Lange <john@y2038.com>

johnlange2 committed:
drivers: modem: Adding header file for SMS messaging
Adding header file for SMS messaging support

Signed-off-by: John Lange <john@y2038.com>

johnlange2 committed:
drivers: modem: Fixing minor whitespace issues
Fixing a few minor whitespace issues while adding SMS messaging

Signed-off-by: John Lange <john@y0238.com>

johnlange2 committed:
drivers: modem: Fixing minor whitespace issues
Fixing a few minor whitespace issues while adding SMS messaging

Signed-off-by: John Lange <john@y0238.com>

johnlange2 committed:
drivers: modem: Adding support for SMS messaging
Adding function pointers for sending and receiving SMS messages
to the modem context.

Signed-off-by: John Lange <john@y2038.com>

johnlange2 committed:
drivers: modem: Adding support for SMS messaging
Adding function pointers for sending and receiving SMS messages
to the modem context.

Signed-off-by: John Lange <john@y2038.com>

johnlange2 committed:
Merge branch 'tmo-add-modem-sms-support' of https://gitlab.com/tmobile/suhwdev/firmware/tmo-zephyr into tmo-add-modem-sms-support

jukkar commented on drivers/modem/Kconfig:
```diff
@@ -139,6 +139,24 @@ config MODEM_CELL_INFO
+
+config MODEM_SMS_OUT_MSG_MAX_LEN
+        int "Maximum length of outbound SMS messages"
+        default 160
```
I suppose the typical scenario is that both the inbound and outbound values are the same. So in order to avoid possible mismatch we could have here `default MODEM_SMS_IN_MSG_MAX_LEN`. Then user could override it if needed, what do you think?

jukkar commented on drivers/modem/modem_sms.h:
```diff
@@ -16,27 +16,19 @@
 struct sms_out {
         char phone[SMS_PHONE_MAX_LEN];
-        char msg  [CONFIG_SMS_OUT_MSG_MAX_LEN];
+        char msg  [CONFIG_MODEM_SMS_OUT_MSG_MAX_LEN + 2];
```
Why +2, wouldn't +1 be enough?

jukkar commented on drivers/modem/Kconfig:
```diff
@@ -139,6 +139,24 @@ config MODEM_CELL_INFO
 	help
```
BTW, we do not do merge branches in zephyr history so please rebase the PR.

johnlange2 replied:
Will do.

johnlange2 commented on drivers/modem/modem_sms.h:
```diff
@@ -0,0 +1,42 @@
+};
+
+typedef int (*send_sms_func_t)(void *obj, const struct sms_out *sms);
+typedef int (*recv_sms_func_t)(void *obj,       struct sms_in  *sms, k_timeout_t timeout);
```
Assuming no objections, I'll add the timeout to struct sms_in

ycsin reviewed (changes requested):
I think not all modem will/can implement this function, i.e. wifi modems, so probably we need a Kconfig to prevent those modem from compiling the sms part (context, shell), or at least perform NULL check so that the shell wont invoke a null function

ycsin commented:
Offtopic
@rlubos @jukkar @rerickson1 do you think if the RTC local time query can be done similarly like this PR?
